File: src/main/java/intellistart/interviewplanning/model/booking/validation/BookingValidator.java
package intellistart.interviewplanning.model.booking.validation;

import intellistart.interviewplanning.exceptions.BookingException;
import intellistart.interviewplanning.exceptions.BookingException.BookingExceptionProfile;
import intellistart.interviewplanning.exceptions.BookingLimitException;
import intellistart.interviewplanning.exceptions.BookingLimitException.BookingLimitExceptionProfile;
import intellistart.interviewplanning.exceptions.SlotException;
import intellistart.interviewplanning.exceptions.SlotException.SlotExceptionProfile;
import intellistart.interviewplanning.exceptions.UserException;
import intellistart.interviewplanning.model.booking.Booking;
import intellistart.interviewplanning.model.bookinglimit.BookingLimitService;
import intellistart.interviewplanning.model.candidateslot.CandidateSlot;
import intellistart.interviewplanning.model.interviewerslot.InterviewerSlot;
import intellistart.interviewplanning.model.interviewerslot.InterviewerSlotService;
import intellistart.interviewplanning.model.period.Period;
import intellistart.interviewplanning.model.period.PeriodService;
import intellistart.interviewplanning.model.period.services.TimeService;
import intellistart.interviewplanning.model.user.User;
import intellistart.interviewplanning.model.week.WeekService;
import java.time.LocalDate;
import java.util.Collection;
import java.util.List;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * Business logic Booking validator.
 */
@Component
public class BookingValidator {

  private static final int BOOKING_PERIOD_DURATION_MINUTES = 90;
  private static final int DESCRIPTION_MAX_SIZE = 4000;
  private static final int SUBJECT_MAX_SIZE = 255;
  private final PeriodService periodService;
  private final TimeService timeService;
  private final WeekService weekService;
  private final BookingLimitService bookingLimitService;
  private final InterviewerSlotService interviewerSlotService;

  /**
   * Constructor.
   */
  @Autowired
  public BookingValidator(PeriodService periodService, TimeService timeService,
      WeekService weekService, BookingLimitService bookingLimitService,
      InterviewerSlotService interviewerSlotService) {
    this.periodService = periodService;
    this.timeService = timeService;
    this.weekService = weekService;
    this.bookingLimitService = bookingLimitService;
    this.interviewerSlotService = interviewerSlotService;
  }

  /**
   * Perform business logic validation for updating Booking.
   *
   * @param updatingBooking Booking with old parameters
   * @param newDataBooking Booking with new parameters
   *
   * @throws UserException          if new booking's interviewer has invalid role
   * @throws SlotException          if duration of new Period is invalid
   * @throws BookingLimitException  if interviewer's booking limit is exceeded
   * @throws BookingException       if periods of InterviewSlot, CandidateSlot do not
   *                                intersect with new Period or new Period is overlapping
   *                                with existing Periods of InterviewerSlot and
   *                                CandidateSlot
   */
  public void validateUpdating(Booking updatingBooking, Booking newDataBooking)
      throws SlotException, BookingException, UserException, BookingLimitException {

    Period newPeriod = newDataBooking.getPeriod();

    int periodDuration = timeService.calculateDurationMinutes(
        newPeriod.getFrom(), newPeriod.getTo());
    if (periodDuration != BOOKING_PERIOD_DURATION_MINUTES) {
      throw new SlotException(SlotExceptionProfile.INVALID_BOUNDARIES);
    }

    if (newDataBooking.getSubject().length() > SUBJECT_MAX_SIZE) {
      throw new BookingException(BookingExceptionProfile.INVALID_SUBJECT);
    }
    if (newDataBooking.getDescription().length() > DESCRIPTION_MAX_SIZE) {
      throw new BookingException(BookingExceptionProfile.INVALID_DESCRIPTION);
    }

    InterviewerSlot newInterviewerSlot = newDataBooking.getInterviewerSlot();
    CandidateSlot newCandidateSlot = newDataBooking.getCandidateSlot();

    LocalDate interviewerSlotDate = weekService.convertToLocalDate(
        newInterviewerSlot.getWeek().getId(), newInterviewerSlot.getDayOfWeek());

    if (!interviewerSlotDate.equals(newCandidateSlot.getDate())) {
      throw new BookingException(BookingExceptionProfile.SLOTS_NOT_INTERSECTING);
    }

    if (!periodService.isFirstInsideSecond(newPeriod, newInterviewerSlot.getPeriod())
        || !periodService.isFirstInsideSecond(newPeriod, newCandidateSlot.getPeriod())) {
      throw new BookingException(BookingExceptionProfile.SLOTS_NOT_INTERSECTING);
    }

    Collection<Booking> interviewSlotBookings = newInterviewerSlot.getBookings();
    Collection<Booking> candidateSlotBookings = newCandidateSlot.getBookings();

    validatePeriodNotOverlappingWithOtherBookingPeriods(
        updatingBooking, newPeriod, interviewSlotBookings);

    validatePeriodNotOverlappingWithOtherBookingPeriods(
        updatingBooking, newPeriod, candidateSlotBookings);

    if (!newInterviewerSlot.equals(updatingBooking.getInterviewerSlot())) {
      User newInterviewer = newInterviewerSlot.getUser();
      List<InterviewerSlot> interviewerSlotsNewInterviewer = interviewerSlotService
          .getInterviewerSlotsByUserAndWeek(
              newInterviewer, newInterviewerSlot.getWeek());

      long bookingsNumber = interviewerSlotsNewInterviewer.stream()
          .map(InterviewerSlot::getBookings)
          .flatMap(Collection::stream)
          .count();

      long bookingLimit = bookingLimitService
          .getBookingLimitByInterviewer(
                  newInterviewer,
                  newInterviewerSlot.getWeek()).getBookingLimit();

      if (bookingsNumber >= bookingLimit) {
        throw new BookingLimitException(BookingLimitExceptionProfile.BOOKING_LIMIT_IS_EXCEEDED);
      }
    }
  }

  private void validatePeriodNotOverlappingWithOtherBookingPeriods(
      Booking updatingBooking, Period period, Collection<Booking> bookings)
      throws BookingException {

    for (Booking booking : bookings) {
      if (periodService.areOverlapping(booking.getPeriod(), period)
          && !booking.equals(updatingBooking)) {
        throw new BookingException(BookingExceptionProfile.SLOTS_NOT_INTERSECTING);
      }
    }
  }

  /**
   * Alias for {@link #validateUpdating(Booking, Booking)}.
   */
  public void validateCreating(Booking newBooking)
      throws SlotException, BookingException, BookingLimitException, UserException {
    validateUpdating(new Booking(), newBooking);
  }
}

File: app/src/main/java/intellistart/interviewplanning/model/booking/validation/BookingValidator.kt
package intellistart.interviewplanning.model.booking.validation

import intellistart.interviewplanning.exceptions.BookingException
import intellistart.interviewplanning.exceptions.BookingException.BookingExceptionProfile
import intellistart.interviewplanning.exceptions.SlotException
import intellistart.interviewplanning.exceptions.SlotException.SlotExceptionProfile
import intellistart.interviewplanning.model.booking.Booking
import intellistart.interviewplanning.model.period.Period
import intellistart.interviewplanning.model.period.PeriodService
import intellistart.interviewplanning.model.period.TimeService
import intellistart.interviewplanning.model.slot.SlotService
import org.bson.types.ObjectId
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono
import reactor.kotlin.core.util.function.component1
import reactor.kotlin.core.util.function.component2

@Component
class BookingValidator(
    private val periodService: PeriodService,
    private val timeService: TimeService,
    private val slotService: SlotService,
) {

    fun validateCreating(newBooking: Booking): Mono<Unit> = validateUpdating(newBooking)

    fun validateUpdating(newBooking: Booking): Mono<Unit> =
        Mono.`when`(
            Mono.fromCallable {
                checkSubjectAndDescriptionLength(newBooking.subject, newBooking.description)
                checkBookingPeriodForNinetyMinutes(newBooking.period)
            },
            checkBookingForOverlapsWithSlots(newBooking),
            checkBookingForOverlapsWithOtherBookings(newBooking),
        ).thenReturn(Unit)

    private fun checkBookingPeriodForNinetyMinutes(period: Period) {
        if (timeService.calculateDurationMinutes(period.from, period.to) < BOOKING_PERIOD_DURATION_MINUTES) {
            throw SlotException(SlotExceptionProfile.INVALID_BOUNDARIES)
        }
    }

    private fun checkSubjectAndDescriptionLength(subject: String, description: String) {
        if (subject.length > SUBJECT_MAX_SIZE) {
            throw BookingException(BookingExceptionProfile.INVALID_SUBJECT)
        }
        if (description.length > DESCRIPTION_MAX_SIZE) {
            throw BookingException(BookingExceptionProfile.INVALID_DESCRIPTION)
        }
    }

    private fun checkBookingForOverlapsWithSlots(booking: Booking): Mono<Unit> = Mono.zip(
        slotService.getById(booking.interviewerSlotId),
        slotService.getById(booking.candidateSlotId)
    ).flatMap { tuple ->
        val periodOfInterviewer: Period = tuple.t1.period
        val periodOfCandidate: Period = tuple.t2.period
        val bookingPeriod: Period = booking.period

        if (
            !periodOfInterviewer.date.isEqual(periodOfCandidate.date) ||
            !periodService.isFirstInsideSecond(bookingPeriod, periodOfInterviewer) ||
            !periodService.isFirstInsideSecond(bookingPeriod, periodOfCandidate)
        ) {
            Mono.error(BookingException(BookingExceptionProfile.SLOTS_NOT_INTERSECTING))
        } else {
            Unit.toMono()
        }
    }

    private fun checkBookingForOverlapsWithOtherBookings(
        booking: Booking,
    ): Mono<Unit> = Mono.zip(
        slotService.getById(booking.interviewerSlotId),
        slotService.getById(booking.candidateSlotId)
    ).handle { (interviewerSlot, candidateSlot), sink ->
        val interviewerBookings: Collection<Booking> = interviewerSlot.bookings
        val candidateBookings: Collection<Booking> = candidateSlot.bookings

        validatePeriodNotOverlappingWithOtherBookingPeriods(
            booking.id,
            booking.period,
            interviewerBookings
        )

        validatePeriodNotOverlappingWithOtherBookingPeriods(
            booking.id,
            booking.period,
            candidateBookings
        )
        sink.complete()
    }

    private fun validatePeriodNotOverlappingWithOtherBookingPeriods(
        updatingBookingId: ObjectId,
        period: Period,
        bookings: Collection<Booking>,
    ) {
        bookings.forEach { booking ->
            if (booking.id != updatingBookingId && periodService.areOverlapping(booking.period, period)) {
                throw BookingException(BookingExceptionProfile.SLOTS_NOT_INTERSECTING)
            }
        }
    }

    companion object {
        const val BOOKING_PERIOD_DURATION_MINUTES = 90
        const val DESCRIPTION_MAX_SIZE = 4000
        const val SUBJECT_MAX_SIZE = 255
    }
}
